wrappers/services.go

```diff
@@ -204,12 +204,56 @@ type StartServicesFlags struct {
 // are services. Service units will be started in the order provided by the
 // caller.
 func StartServices(apps []*snap.AppInfo, disabledSvcs []string, flags *StartServicesFlags, inter interacter, tm timings.Measurer) (err error) {
+	if flags == nil {
+		flags = &StartServicesFlags{}
+	}
 	systemSysd := systemd.New(dirs.GlobalRootDir, systemd.SystemMode, inter)
 	userSysd := systemd.New(dirs.GlobalRootDir, systemd.GlobalUserMode, inter)
 	cli := client.New()
 
+	var disableEnabledServices func()
+
+	defer func() {
+		if err == nil {
+			return
+		}
+		if disableEnabledServices != nil {
+			disableEnabledServices()
+		}
+	}()
+
+	var toEnable []*snap.AppInfo
 	systemServices := make([]string, 0, len(apps))
 	userServices := make([]string, 0, len(apps))
+
+	for _, app := range apps {
+		// they're *supposed* to be all services, but checking doesn't hurt
+		if !app.IsService() {
+			continue
+		}
+		// sockets and timers are enabled and started separately (and unconditionally) further down.
+		if len(app.Sockets) == 0 && app.Timer == nil {
+			if strutil.ListContains(disabledSvcs, app.Name) {
+				continue
+			}
+			svcName := app.ServiceName()
+			switch app.DaemonScope {
+			case snap.SystemDaemon:
+				systemServices = append(systemServices, svcName)
+			case snap.UserDaemon:
+				userServices = append(userServices, svcName)
+			}
+			if flags.Enable {
+				toEnable = append(toEnable, app)
+			}
+		}
+	}
+
+	disableEnabledServices, err = enableServices(toEnable, inter)
+	if err != nil {
+		return err
+	}
+
 	for _, app := range apps {
```

Comment:
this loop could use a comment explaining that this loop handles sockets and timers